lib/search_query_builder.rb

```diff
@@ -0,0 +1,177 @@
+        't' => 'card_type_id',
+        'text' => 'stripped_text',
+        'title' => 'stripped_title',
+        'trash_cost' => 'trash_cost',
+        'u' => 'is_unique',
+        'v' => 'agenda_points',
+        'x' => 'stripped_text',
+    }
+
+    def initialize(query)
+        @query = query
+        @parse_error = nil
+        @parse_tree = nil
+        @where = ''
+        @where_values = []
+        begin
+            @parse_tree = @@parser.parse(@query)
+        rescue Parslet::ParseFailed => e
+            @parse_error = e
+        end
+        if @parse_error != nil
+            return
+        end
+        constraints = []
+        where = []
+        @parse_tree[:fragments].each {|f|
+            if f.include?(:search_term)
+                keyword = f[:search_term][:keyword].to_s
+                match_type = f[:search_term][:match_type].to_s
+                value = f[:search_term][:value][:string].to_s
+                if @@boolean_keywords.include?(keyword)
+                    operator = ''
+                    if @@boolean_operators.include?(match_type)
+                        operator = @@boolean_operators[match_type]
+                    else
+                        @parse_error = 'Invalid boolean operator "%s"' % match_type
+                        return
+                    end
+                    constraints << '%s %s ?' % [@@term_to_field_map[keyword], operator]
+                    where << value
+                elsif @@numeric_keywords.include?(keyword)
+                    operator = ''
+                    if @@numeric_operators.include?(match_type)
+                        operator = @@numeric_operators[match_type]
+                    else
+                        @parse_error = 'Invalid numeric operator "%s"' % match_type
+                        return
+                    end
+                    constraints << '%s %s ?' % [@@term_to_field_map[keyword], operator]
+                    where << value 
+                else
+                    # String fields only support : and !
+                    operator = ''
+                    if @@string_operators.include?(match_type)
+                        operator = @@string_operators[match_type]
+                    else
+                        @parse_error = 'Invalid string operator "%s"' % match_type
+                        return
+                    end
+                    constraints << 'lower(%s) %s ?' % [@@term_to_field_map[keyword], operator]
+                    where << '%%%s%%' % value
+                end
+            end
+
+            # bare/quoted words in the query are automatically mapped to stripped_title
+            if f.include?(:string)
+                    value = f[:string].to_s
+                    operator = value.start_with?('!') ? 'NOT LIKE' : 'LIKE'
+                    value    = value.start_with?('!') ? value[1..] : value
+                    constraints << 'lower(stripped_title) %s ?' % operator
```

Comment:
Is `lower` correct here? Will it match if given a capitalized term?

Reply:
not without altering the input values, which i've done now.  Thanks for the nudge.